yamllint/rules/new_lines.py

```diff
@@ -47,7 +52,12 @@ def check(conf, line):
                     line.buffer[line.end:line.end + 2] != '\r\n'):
                 yield LintProblem(1, line.end - line.start + 1,
                                   'wrong new line character: expected \\r\\n')
-        else:
+        elif conf['type'] == 'unix':
             if line.buffer[line.end] != '\n':
                 yield LintProblem(1, line.end - line.start + 1,
                                   'wrong new line character: expected \\n')
+        elif conf['type'] == 'auto':
+            if line.buffer[line.end:line.end + len(os_linesep)] != os_linesep:
+                yield LintProblem(1, line.end - line.start + 1,
+                                  'wrong new line character: expected {}'
+                                  .format(repr(os_linesep)))
```

Comment:
I propose something more performant, that doesn't induce code repetition: don't touch this part, but just add right after `def check(conf, line):`:
```python
if conf['type'] == 'platform':
    conf['type'] = 'dos' if linesep == '\r\n' else 'unix'
```
What do you think?

Reply:
I also thought about this option, but the current implementations would allow for any type of `linesep` (even if there are probably no others...). Also should windows ever decide to switch to `\n` this part would still work. 

Maybe changing the order of the `if`, `elif`, blocks would be the better way to improve performance:

```python
if conf['type'] == 'unix':  # most performant, only 1 if check, probably most used
    ...
elif conf['type'] == 'platform':  # will probably become second most used, 2 if check
    ...
elif conf['type'] == 'dos':  # will probably be used the least, 2 if checks
    ...
```

Checking for the platform first would intreduce two if statements in any case, but I don't know if branch prediction helps here enough to make this the better option so.

I also find it a little more readable to have if, elif blocks, but thats personal preference. Also the code is already duplicated, so one more duplication doesn't hurt to badly in my opinion.

Your call, whatever you prefere.

Reply:
I would recomend this how ever:

```python
if conf['type'] == 'platform':
    conf['type'] = 'dos' if sys.platform == 'win32' else 'unix'
```

Reply:
I just thought of another alternative that reduces code duplication by a lot:

 ```python
def check(conf, line):
    if conf['type'] == 'unix':
        newline_char = '\n'
    elif conf['type'] == 'platform':
        newline_char = linesep
    elif conf['type'] == 'dos':
        newline_char = '\r\n'

    print(len(line.buffer))
    if line.start == 0 and len(line.buffer) > line.end:
        if line.buffer[line.end:line.end + len(newline_char)] != newline_char:
            yield LintProblem(1, line.end - line.start + 1,
                                'wrong new line character: expected {}'
                                .format(repr(newline_char).strip('\'')))
```

what do you think about that?

---

potential Problem: this looses the line-lenth check on dos systems:

https://github.com/adrienverge/yamllint/blob/632665c3e68b0743a98035e0f173a7ad85e37c4f/yamllint/rules/new_lines.py#L46

I am not sure what exacly this is guarding aganinst? I assume files that end in `\n` might result in an IndexError? or am I overlooking something?

Reply:
> I propose:
> ```
> -if line.buffer[line.end:line.end + len(newline_char)] != newline_char:
> +if line.buffer[line.end:] != newline_char:
> ```

this was also my fist idea, but doesn't work because `line.buffer` containes all lines...

Reply:
Argh...